File: src/sttp/sttp-core/Storage/InMemoryNodeStore.cs
using SttpMcp.Domain.Contracts;
using SttpMcp.Domain.Models;
using System.Text.Json;

namespace SttpMcp.Storage;

public sealed class InMemoryNodeStore : INodeStore, INodeStoreInitializer
{
    private const string DefaultTenantId = "default";
    private const string TenantScopePrefix = "tenant:";
    private const string TenantScopeSeparator = "::session:";

    private readonly List<SttpNode> _nodes = [];
    private readonly List<(string SessionId, AvecState Avec, string Trigger)> _calibrations = [];
    private readonly List<SyncCheckpoint> _checkpoints = [];

    public Task InitializeAsync(CancellationToken ct = default) => Task.CompletedTask;

    public Task<IReadOnlyList<SttpNode>> QueryNodesAsync(NodeQuery query, CancellationToken ct = default)
    {
        var cappedLimit = Math.Max(1, query.Limit);

        var result = _nodes
            .Where(n => string.IsNullOrWhiteSpace(query.SessionId) || n.SessionId == query.SessionId)
            .Where(n => query.FromUtc is null || n.Timestamp >= query.FromUtc.Value)
            .Where(n => query.ToUtc is null || n.Timestamp <= query.ToUtc.Value)
            .OrderByDescending(n => n.Timestamp)
            .Take(cappedLimit)
            .ToList();

        return Task.FromResult<IReadOnlyList<SttpNode>>(result);
    }

    public Task<NodeUpsertResult> UpsertNodeAsync(SttpNode node, CancellationToken ct = default)
    {
        var syncKey = string.IsNullOrWhiteSpace(node.SyncKey)
            ? node.CanonicalSyncKey()
            : node.SyncKey.Trim();
        var now = DateTime.UtcNow;
        var candidate = node with
        {
            SyncKey = syncKey,
            UpdatedAt = now
        };

        var existingIndex = _nodes.FindIndex(existing =>
            existing.SessionId == candidate.SessionId &&
            existing.SyncKey == syncKey);

        if (existingIndex >= 0)
        {
            var existing = _nodes[existingIndex];
            if (!AreEquivalent(existing, candidate))
            {
                _nodes[existingIndex] = candidate;
                return Task.FromResult(new NodeUpsertResult
                {
                    NodeId = GetNodeId(existing),
                    SyncKey = syncKey,
                    Status = NodeUpsertStatus.Updated,
                    UpdatedAt = now
                });
            }

            return Task.FromResult(new NodeUpsertResult
            {
                NodeId = GetNodeId(existing),
                SyncKey = syncKey,
                Status = NodeUpsertStatus.Duplicate,
                UpdatedAt = existing.UpdatedAt
            });
        }

        _nodes.Add(candidate);
        return Task.FromResult(new NodeUpsertResult
        {
            NodeId = GetNodeId(candidate),
            SyncKey = syncKey,
            Status = NodeUpsertStatus.Created,
            UpdatedAt = now
        });
    }

    public Task<IReadOnlyList<SttpNode>> GetByResonanceAsync(
        string sessionId, AvecState current, int limit = 5, CancellationToken ct = default)
    {
        var result = _nodes
            .Where(n => n.SessionId == sessionId)
            .OrderBy(n => Math.Abs(n.Psi - current.Psi))
            .Take(limit)
            .ToList();

        return Task.FromResult<IReadOnlyList<SttpNode>>(result);
    }

    public Task<IReadOnlyList<SttpNode>> ListNodesAsync(
        int limit = 50,
        string? sessionId = null,
        CancellationToken ct = default)
        => QueryNodesAsync(new NodeQuery { Limit = Math.Clamp(limit, 1, 200), SessionId = sessionId }, ct);

    public Task<AvecState?> GetLastAvecAsync(
        string sessionId, CancellationToken ct = default)
    {
        var last = _calibrations
            .Where(c => c.SessionId == sessionId)
            .Select(c => c.Avec)
            .LastOrDefault();

        return Task.FromResult<AvecState?>(last);
    }

    public Task<IReadOnlyList<string>> GetTriggerHistoryAsync(
        string sessionId, CancellationToken ct = default)
    {
        var history = _calibrations
            .Where(c => c.SessionId == sessionId)
            .Select(c => c.Trigger)
            .ToList();

        return Task.FromResult<IReadOnlyList<string>>(history);
    }

    public Task StoreCalibrationAsync(
        string sessionId, AvecState avec, string trigger, CancellationToken ct = default)
    {
        _calibrations.Add((sessionId, avec, trigger));
        return Task.CompletedTask;
    }

    public Task<ChangeQueryResult> QueryChangesSinceAsync(
        string sessionId,
        SyncCursor? cursor,
        int limit,
        CancellationToken ct = default)
    {
        var cappedLimit = Math.Max(1, limit);
        var changes = _nodes
            .Where(node => node.SessionId == sessionId)
            .Where(node => cursor is null ||
                node.UpdatedAt > cursor.UpdatedAt ||
                (node.UpdatedAt == cursor.UpdatedAt && string.CompareOrdinal(node.SyncKey, cursor.SyncKey) > 0))
            .OrderBy(node => node.UpdatedAt)
            .ThenBy(node => node.SyncKey, StringComparer.Ordinal)
            .ToList();

        var hasMore = changes.Count > cappedLimit;
        if (hasMore)
            changes = changes.Take(cappedLimit).ToList();

        return Task.FromResult(new ChangeQueryResult
        {
            Nodes = changes,
            HasMore = hasMore,
            NextCursor = changes.Count == 0
                ? null
                : new SyncCursor
                {
                    UpdatedAt = changes[^1].UpdatedAt,
                    SyncKey = changes[^1].SyncKey
                }
        });
    }

    public Task<SyncCheckpoint?> GetCheckpointAsync(
        string sessionId,
        string connectorId,
        CancellationToken ct = default)
    {
        var checkpoint = _checkpoints.FirstOrDefault(existing =>
            existing.SessionId == sessionId &&
            existing.ConnectorId == connectorId);
        return Task.FromResult(checkpoint);
    }

    public Task PutCheckpointAsync(SyncCheckpoint checkpoint, CancellationToken ct = default)
    {
        var existingIndex = _checkpoints.FindIndex(existing =>
            existing.SessionId == checkpoint.SessionId &&
            existing.ConnectorId == checkpoint.ConnectorId);

        if (existingIndex >= 0)
            _checkpoints[existingIndex] = checkpoint;
        else
            _checkpoints.Add(checkpoint);

        return Task.CompletedTask;
    }

    public Task<BatchRekeyResult> BatchRekeyScopesAsync(
        IReadOnlyList<string> nodeIds,
        string targetTenantId,
        string targetSessionId,
        bool dryRun,
        bool allowMerge,
        CancellationToken ct = default)
    {
        if (nodeIds.Count == 0)
            throw new ArgumentException("at least one node id is required", nameof(nodeIds));

        if (string.IsNullOrWhiteSpace(targetSessionId))
            throw new ArgumentException("target session id cannot be empty", nameof(targetSessionId));

        var normalizedTargetTenantId = NormalizeTenantId(targetTenantId);
        var normalizedNodeIds = nodeIds
            .Select(NormalizeTemporalNodeId)
            .Where(nodeId => !string.IsNullOrWhiteSpace(nodeId))
            .Distinct(StringComparer.Ordinal)
            .ToList();

        if (normalizedNodeIds.Count == 0)
            throw new ArgumentException("no valid node ids were provided", nameof(nodeIds));

        var missingNodeIds = new List<string>();
        var scopeKeys = new SortedSet<(string TenantId, string SessionId)>();

        foreach (var nodeId in normalizedNodeIds)
        {
            var anchor = _nodes.FirstOrDefault(node => string.Equals(GetNodeId(node), nodeId, StringComparison.Ordinal));
            if (anchor is null || string.IsNullOrWhiteSpace(anchor.SessionId))
            {
                missingNodeIds.Add(nodeId!);
                continue;
            }

            scopeKeys.Add((DeriveTenantIdFromSession(anchor.SessionId), anchor.SessionId));
        }

        var scopeResults = new List<ScopeRekeyResult>();
        var temporalNodesUpdated = 0;
        var calibrationsUpdated = 0;

        foreach (var scope in scopeKeys)
        {
            var temporalNodes = _nodes.Count(node =>
                node.SessionId == scope.SessionId &&
                MatchesTenant(node.SessionId, scope.TenantId));
            var calibrations = _calibrations.Count(calibration =>
                calibration.SessionId == scope.SessionId &&
                MatchesTenant(calibration.SessionId, scope.TenantId));

            var sameScope = scope.TenantId == normalizedTargetTenantId && scope.SessionId == targetSessionId;
            var targetTemporalNodes = sameScope
                ? 0
                : _nodes.Count(node =>
                    node.SessionId == targetSessionId &&
                    MatchesTenant(node.SessionId, normalizedTargetTenantId));
            var targetCalibrations = sameScope
                ? 0
                : _calibrations.Count(calibration =>
                    calibration.SessionId == targetSessionId &&
                    MatchesTenant(calibration.SessionId, normalizedTargetTenantId));

            var conflict = !allowMerge && !sameScope && (targetTemporalNodes > 0 || targetCalibrations > 0);
            var applied = false;
            string? message = null;

            if (sameScope)
            {
                message = "source and target scopes are identical";
            }
            else if (conflict)
            {
                message = "target scope already contains rows; set allowMerge=true to override";
            }
            else if (!dryRun)
            {
                for (var index = 0; index < _nodes.Count; index++)
                {
                    var node = _nodes[index];
                    if (node.SessionId == scope.SessionId && MatchesTenant(node.SessionId, scope.TenantId))
                        _nodes[index] = node with { SessionId = targetSessionId };
                }

                for (var index = 0; index < _calibrations.Count; index++)
                {
                    var calibration = _calibrations[index];
                    if (calibration.SessionId == scope.SessionId && MatchesTenant(calibration.SessionId, scope.TenantId))
                        _calibrations[index] = (targetSessionId, calibration.Avec, calibration.Trigger);
                }

                applied = true;
                temporalNodesUpdated += temporalNodes;
                calibrationsUpdated += calibrations;
            }

            scopeResults.Add(new ScopeRekeyResult
            {
                SourceTenantId = scope.TenantId,
                SourceSessionId = scope.SessionId,
                TargetTenantId = normalizedTargetTenantId,
                TargetSessionId = targetSessionId,
                TemporalNodes = temporalNodes,
                Calibrations = calibrations,
                TargetTemporalNodes = targetTemporalNodes,
                TargetCalibrations = targetCalibrations,
                Applied = applied,
                Conflict = conflict,
                Message = message
            });
        }

        return Task.FromResult(new BatchRekeyResult
        {
            DryRun = dryRun,
            RequestedNodeIds = normalizedNodeIds.Count,
            ResolvedNodeIds = normalizedNodeIds.Count - missingNodeIds.Count,
            MissingNodeIds = missingNodeIds,
            Scopes = scopeResults,
            TemporalNodesUpdated = temporalNodesUpdated,
            CalibrationsUpdated = calibrationsUpdated
        });
    }

    private static bool AreEquivalent(SttpNode left, SttpNode right)
        => left.Raw == right.Raw
            && left.SessionId == right.SessionId
            && left.Tier == right.Tier
            && left.Timestamp == right.Timestamp
            && left.CompressionDepth == right.CompressionDepth
            && left.ParentNodeId == right.ParentNodeId
            && left.SyncKey == right.SyncKey
            && left.UserAvec == right.UserAvec
            && left.ModelAvec == right.ModelAvec
            && left.CompressionAvec == right.CompressionAvec
            && left.Rho == right.Rho
            && left.Kappa == right.Kappa
            && left.Psi == right.Psi
            && NormalizeMetadata(left.SourceMetadata) == NormalizeMetadata(right.SourceMetadata);

    private static string GetNodeId(SttpNode node)
        => string.IsNullOrWhiteSpace(node.SyncKey)
            ? node.CanonicalSyncKey()
            : node.SyncKey;

    private static string? NormalizeMetadata(ConnectorMetadata? metadata)
        => metadata is null ? null : JsonSerializer.Serialize(metadata);

    private static string DeriveTenantIdFromSession(string sessionId)
    {
        if (!sessionId.StartsWith(TenantScopePrefix, StringComparison.Ordinal))
            return DefaultTenantId;

        var remainder = sessionId[TenantScopePrefix.Length..];
        var separatorIndex = remainder.IndexOf(TenantScopeSeparator, StringComparison.Ordinal);
        if (separatorIndex <= 0)
            return DefaultTenantId;

        var tenantId = remainder[..separatorIndex].Trim();
        return string.IsNullOrWhiteSpace(tenantId) ? DefaultTenantId : tenantId;
    }

    private static string NormalizeTenantId(string? tenantId)
        => string.IsNullOrWhiteSpace(tenantId) ? DefaultTenantId : tenantId.Trim();

    private static bool MatchesTenant(string sessionId, string tenantId)
        => string.Equals(DeriveTenantIdFromSession(sessionId), tenantId, StringComparison.Ordinal);

    private static string? NormalizeTemporalNodeId(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
            return null;

        var trimmed = value.Trim();
        if (!trimmed.StartsWith("temporal_node:", StringComparison.Ordinal))
            return trimmed;

        var normalized = trimmed["temporal_node:".Length..].Trim();
        return string.IsNullOrWhiteSpace(normalized) ? null : normalized;
    }
}

File: src/sttp/sttp-core-rs/src/storage/in_memory_node_store.rs
use anyhow::{Result, anyhow};
use async_trait::async_trait;
use chrono::Utc;
use serde_json::Value;
use tokio::sync::RwLock;
use uuid::Uuid;

use crate::domain::contracts::{NodeStore, NodeStoreInitializer};
use crate::domain::models::{
    AvecState, BatchRekeyResult, ChangeQueryResult, NodeQuery, NodeUpsertResult,
    NodeUpsertStatus, ScopeRekeyResult, SttpNode, SyncCheckpoint, SyncCursor,
};

const DEFAULT_TENANT: &str = "default";
const TENANT_SCOPE_PREFIX: &str = "tenant:";
const TENANT_SCOPE_SEPARATOR: &str = "::session:";

#[derive(Debug, Clone, Copy)]
struct CalibrationRecord {
    avec: AvecState,
}

#[derive(Debug, Default)]
pub struct InMemoryNodeStore {
    nodes: RwLock<Vec<(String, SttpNode)>>,
    calibrations: RwLock<Vec<(String, CalibrationRecord, String)>>,
    checkpoints: RwLock<Vec<SyncCheckpoint>>,
}

impl InMemoryNodeStore {
    pub fn new() -> Self {
        Self::default()
    }
}

#[async_trait]
impl NodeStoreInitializer for InMemoryNodeStore {
    async fn initialize_async(&self) -> Result<()> {
        Ok(())
    }
}

#[async_trait]
impl NodeStore for InMemoryNodeStore {
    async fn query_nodes_async(&self, query: NodeQuery) -> Result<Vec<SttpNode>> {
        let capped_limit = query.limit.max(1);
        let nodes = self.nodes.read().await;

        let mut result = nodes
            .iter()
            .map(|(_, node)| node)
            .filter(|n| {
                query
                    .session_id
                    .as_ref()
                    .map(|s| &n.session_id == s)
                    .unwrap_or(true)
            })
            .filter(|n| query.from_utc.map(|from| n.timestamp >= from).unwrap_or(true))
            .filter(|n| query.to_utc.map(|to| n.timestamp <= to).unwrap_or(true))
            .cloned()
            .collect::<Vec<_>>();

        result.sort_by(|a, b| b.timestamp.cmp(&a.timestamp));
        result.truncate(capped_limit);

        Ok(result)
    }

    async fn upsert_node_async(&self, node: SttpNode) -> Result<NodeUpsertResult> {
        let mut candidate = node;
        let sync_key = if candidate.sync_key.trim().is_empty() {
            candidate.canonical_sync_key()
        } else {
            candidate.sync_key.trim().to_string()
        };

        let now = Utc::now();
        candidate.sync_key = sync_key.clone();
        candidate.updated_at = now;

        let mut nodes = self.nodes.write().await;
        if let Some((node_id, existing)) = nodes.iter_mut().find(|(_, existing)| {
            existing.session_id == candidate.session_id && existing.sync_key == sync_key
        }) {
            let existing_metadata = normalize_metadata(existing.source_metadata.as_ref());
            let candidate_metadata = normalize_metadata(candidate.source_metadata.as_ref());

            if existing.raw != candidate.raw
                || existing.tier != candidate.tier
                || existing.timestamp != candidate.timestamp
                || existing.compression_depth != candidate.compression_depth
                || existing.parent_node_id != candidate.parent_node_id
                || existing.user_avec != candidate.user_avec
                || existing.model_avec != candidate.model_avec
                || existing.compression_avec != candidate.compression_avec
                || (existing.rho - candidate.rho).abs() > f32::EPSILON
                || (existing.kappa - candidate.kappa).abs() > f32::EPSILON
                || (existing.psi - candidate.psi).abs() > f32::EPSILON
                || existing_metadata != candidate_metadata
            {
                *existing = candidate.clone();
                existing.updated_at = now;
                return Ok(NodeUpsertResult {
                    node_id: node_id.clone(),
                    sync_key,
                    status: NodeUpsertStatus::Updated,
                    updated_at: now,
                });
            }

            return Ok(NodeUpsertResult {
                node_id: node_id.clone(),
                sync_key,
                status: NodeUpsertStatus::Duplicate,
                updated_at: existing.updated_at,
            });
        }

        let node_id = Uuid::new_v4().to_string();
        nodes.push((node_id.clone(), candidate));

        Ok(NodeUpsertResult {
            node_id,
            sync_key,
            status: NodeUpsertStatus::Created,
            updated_at: now,
        })
    }

    async fn get_by_resonance_async(
        &self,
        session_id: &str,
        current_avec: AvecState,
        limit: usize,
    ) -> Result<Vec<SttpNode>> {
        let nodes = self.nodes.read().await;

        let mut result = nodes
            .iter()
            .map(|(_, node)| node)
            .filter(|n| n.session_id == session_id)
            .cloned()
            .collect::<Vec<_>>();

        result.sort_by(|a, b| {
            let ad = (a.psi - current_avec.psi()).abs();
            let bd = (b.psi - current_avec.psi()).abs();
            ad.total_cmp(&bd)
        });
        result.truncate(limit);

        Ok(result)
    }

    async fn list_nodes_async(&self, limit: usize, session_id: Option<&str>) -> Result<Vec<SttpNode>> {
        self.query_nodes_async(NodeQuery {
            limit: limit.clamp(1, 200),
            session_id: session_id.map(|s| s.to_string()),
            from_utc: None,
            to_utc: None,
        })
        .await
    }

    async fn get_last_avec_async(&self, session_id: &str) -> Result<Option<AvecState>> {
        let calibrations = self.calibrations.read().await;
        let last = calibrations
            .iter()
            .rev()
            .find(|(sid, _, _)| sid == session_id)
            .map(|(_, record, _)| record.avec);
        Ok(last)
    }

    async fn get_trigger_history_async(&self, session_id: &str) -> Result<Vec<String>> {
        let calibrations = self.calibrations.read().await;
        let history = calibrations
            .iter()
            .filter(|(sid, _, _)| sid == session_id)
            .map(|(_, _, trigger)| trigger.clone())
            .collect::<Vec<_>>();

        Ok(history)
    }

    async fn store_calibration_async(
        &self,
        session_id: &str,
        avec: AvecState,
        trigger: &str,
    ) -> Result<()> {
        let mut calibrations = self.calibrations.write().await;
        calibrations.push((
            session_id.to_string(),
            CalibrationRecord { avec },
            trigger.to_string(),
        ));
        Ok(())
    }

    async fn query_changes_since_async(
        &self,
        session_id: &str,
        cursor: Option<SyncCursor>,
        limit: usize,
    ) -> Result<ChangeQueryResult> {
        let capped_limit = limit.max(1);
        let nodes = self.nodes.read().await;

        let mut changes = nodes
            .iter()
            .map(|(_, node)| node)
            .filter(|node| node.session_id == session_id)
            .filter(|node| match cursor.as_ref() {
                Some(cursor) => {
                    node.updated_at > cursor.updated_at
                        || (node.updated_at == cursor.updated_at && node.sync_key > cursor.sync_key)
                }
                None => true,
            })
            .cloned()
            .collect::<Vec<_>>();

        changes.sort_by(|left, right| {
            left.updated_at
                .cmp(&right.updated_at)
                .then_with(|| left.sync_key.cmp(&right.sync_key))
        });

        let has_more = changes.len() > capped_limit;
        if has_more {
            changes.truncate(capped_limit);
        }

        let next_cursor = changes.last().map(|node| SyncCursor {
            updated_at: node.updated_at,
            sync_key: node.sync_key.clone(),
        });

        Ok(ChangeQueryResult {
            nodes: changes,
            next_cursor,
            has_more,
        })
    }

    async fn get_checkpoint_async(
        &self,
        session_id: &str,
        connector_id: &str,
    ) -> Result<Option<SyncCheckpoint>> {
        let checkpoints = self.checkpoints.read().await;
        Ok(checkpoints
            .iter()
            .find(|checkpoint| {
                checkpoint.session_id == session_id && checkpoint.connector_id == connector_id
            })
            .cloned())
    }

    async fn put_checkpoint_async(&self, checkpoint: SyncCheckpoint) -> Result<()> {
        let mut checkpoints = self.checkpoints.write().await;
        if let Some(existing) = checkpoints.iter_mut().find(|existing| {
            existing.session_id == checkpoint.session_id
                && existing.connector_id == checkpoint.connector_id
        }) {
            *existing = checkpoint;
        } else {
            checkpoints.push(checkpoint);
        }

        Ok(())
    }

    async fn batch_rekey_scopes_async(
        &self,
        node_ids: Vec<String>,
        target_tenant_id: &str,
        target_session_id: &str,
        dry_run: bool,
        allow_merge: bool,
    ) -> Result<BatchRekeyResult> {
        if node_ids.is_empty() {
            return Err(anyhow!("at least one node id is required"));
        }

        let target_tenant_id = normalize_tenant_id(target_tenant_id);
        if target_session_id.trim().is_empty() {
            return Err(anyhow!("target session id cannot be empty"));
        }

        let normalized_node_ids = node_ids
            .into_iter()
            .filter_map(|node_id| normalize_temporal_node_id(&node_id))
            .collect::<Vec<_>>();

        if normalized_node_ids.is_empty() {
            return Err(anyhow!("no valid node ids were provided"));
        }

        let nodes_snapshot = self.nodes.read().await.clone();
        let calibrations_snapshot = self.calibrations.read().await.clone();

        let mut missing_node_ids = Vec::new();
        let mut scopes = Vec::new();

        for node_id in &normalized_node_ids {
            let Some((_, node)) = nodes_snapshot.iter().find(|(id, _)| id == node_id) else {
                missing_node_ids.push(node_id.clone());
                continue;
            };

            let tenant_id = derive_tenant_id_from_session(&node.session_id);
            if !scopes
                .iter()
                .any(|(tenant, session): &(String, String)| tenant == &tenant_id && session == &node.session_id)
            {
                scopes.push((tenant_id, node.session_id.clone()));
            }
        }

        let mut scope_results = Vec::new();
        let mut temporal_nodes_updated = 0usize;
        let mut calibrations_updated = 0usize;
        let mut sources_to_apply = Vec::new();

        for (source_tenant_id, source_session_id) in scopes {
            let same_scope = source_tenant_id == target_tenant_id && source_session_id == target_session_id;

            let temporal_nodes = nodes_snapshot
                .iter()
                .filter(|(_, node)| node.session_id == source_session_id)
                .count();
            let calibrations = calibrations_snapshot
                .iter()
                .filter(|(session_id, _, _)| session_id == &source_session_id)
                .count();

            let target_temporal_nodes = if same_scope {
                0
            } else {
                nodes_snapshot
                    .iter()
                    .filter(|(_, node)| node.session_id == target_session_id)
                    .count()
            };
            let target_calibrations = if same_scope {
                0
            } else {
                calibrations_snapshot
                    .iter()
                    .filter(|(session_id, _, _)| session_id == target_session_id)
                    .count()
            };

            let conflict = !allow_merge
                && !same_scope
                && (target_temporal_nodes > 0 || target_calibrations > 0);

            let mut applied = false;
            let message = if same_scope {
                Some("source and target scopes are identical".to_string())
            } else if conflict {
                Some("target scope already contains rows; set allow_merge=true to override"
                    .to_string())
            } else {
                if !dry_run {
                    applied = true;
                    temporal_nodes_updated += temporal_nodes;
                    calibrations_updated += calibrations;
                    sources_to_apply.push(source_session_id.clone());
                }
                None
            };

            scope_results.push(ScopeRekeyResult {
                source_tenant_id,
                source_session_id,
                target_tenant_id: target_tenant_id.clone(),
                target_session_id: target_session_id.to_string(),
                temporal_nodes,
                calibrations,
                target_temporal_nodes,
                target_calibrations,
                applied,
                conflict,
                message,
            });
        }

        if !dry_run && !sources_to_apply.is_empty() {
            let mut nodes = self.nodes.write().await;
            for (_, node) in nodes.iter_mut() {
                if sources_to_apply.iter().any(|source| source == &node.session_id) {
                    node.session_id = target_session_id.to_string();
                }
            }

            let mut calibrations = self.calibrations.write().await;
            for (session_id, _, _) in calibrations.iter_mut() {
                if sources_to_apply.iter().any(|source| source == session_id) {
                    *session_id = target_session_id.to_string();
                }
            }
        }

        Ok(BatchRekeyResult {
            dry_run,
            requested_node_ids: normalized_node_ids.len(),
            resolved_node_ids: normalized_node_ids.len().saturating_sub(missing_node_ids.len()),
            missing_node_ids,
            scopes: scope_results,
            temporal_nodes_updated,
            calibrations_updated,
        })
    }
}

fn parse_scoped_session_id(session_id: &str) -> Option<(&str, &str)> {
    let remainder = session_id.strip_prefix(TENANT_SCOPE_PREFIX)?;
    remainder.split_once(TENANT_SCOPE_SEPARATOR)
}

fn derive_tenant_id_from_session(session_id: &str) -> String {
    parse_scoped_session_id(session_id)
        .map(|(tenant, _)| tenant)
        .filter(|tenant| !tenant.trim().is_empty())
        .unwrap_or(DEFAULT_TENANT)
        .to_string()
}

fn normalize_tenant_id(value: &str) -> String {
    let trimmed = value.trim();
    if trimmed.is_empty() {
        DEFAULT_TENANT.to_string()
    } else {
        trimmed.to_string()
    }
}

fn normalize_temporal_node_id(value: &str) -> Option<String> {
    let trimmed = value.trim();
    if trimmed.is_empty() {
        return None;
    }

    if let Some((table, id)) = trimmed.split_once(':') {
        if table != "temporal_node" {
            return None;
        }

        let id = id.trim();
        if id.is_empty() {
            None
        } else {
            Some(id.to_string())
        }
    } else {
        Some(trimmed.to_string())
    }
}

fn normalize_metadata(metadata: Option<&Value>) -> Option<String> {
    metadata.and_then(|value| serde_json::to_string(value).ok())
}
